### PR TITLE
Add executor to execute schedule-plan file

### PR DESCRIPTION
**Pull request by Binyang2014**
Add executor to execute the JSON schedule file generated by msccl-tools

### CODE REVIEW

File: src/include/execution_common.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#define MSCCLPP_EXECUTION_COMMON_HPP_
+
+#include <mscclpp/proxy_channel.hpp>
+#include <mscclpp/sm_channel.hpp>
+
+namespace mscclpp {
+
+constexpr int MAX_CHANNEL = 16;
+constexpr int MAX_CHANNEL_PER_OPERATION = 8;
+constexpr int MAX_OPERATION = 64;
```
Comment (chhwang) [inline src/include/execution_common.hpp:14]: Can we make these configurable?

File: test/executor_test.cc
Comment (chhwang) [inline test/executor_test.cc:1]: Can we make it as a part of `mp_unit_test`?

**Review by chhwang — CHANGES_REQUESTED** [review]

File: src/executor/execution_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "execution_kernel.hpp"
+
+#if defined(MSCCLPP_DEVICE_CUDA)
```
[the remark this answers is not in the document]
Comment (Binyang2014) [reply src/executor/execution_kernel.cu:6]: It doesn't work on AMD, error report is: `ld.lld: error: undefined reference due to --no-allow-shlib-undefined: void mscclpp::ExecutionKernel::launchKernel<mscclpp::LL16Packet>`. Not sure if it is caused by we force to let `CXX=hipcc`

File: include/mscclpp/executor.hpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ private:
+  struct Impl;
+  std::shared_ptr<Impl> impl_;
+
+  friend class Executor;
+};
+
+class Executor {
+ public:
+  Executor(std::shared_ptr<Communicator> comm, int nranksPerNode);
```
[the remark this answers is not in the document]
Comment (Binyang2014) [reply include/mscclpp/executor.hpp:39]: Good point, will change this

File: src/include/execution_common.hpp (hunk 1 above)
Comment (Binyang2014) [reply src/include/execution_common.hpp:14]: I think we'd better to make this config static. It used by the structure `Operation`. And the kernel code uses this structure to execute the generated plan. Static structure would be more efficient for execution.

File: test/executor_test.cc
Comment (Binyang2014) [reply test/executor_test.cc:1]: Fixed

File: include/mscclpp/packet_device.hpp [hunk 4]
```diff
@@ -24,12 +24,20 @@ union alignas(16) LL16Packet {
     uint32_t data2;
     uint32_t flag2;
   };
+  using ValueType = uint2;
```
[the remark this answers is not in the document]
Comment (Binyang2014) [reply include/mscclpp/packet_device.hpp]: Change to `Payload`

**Review by chhwang — APPROVED** [review]
LGTM

File: src/include/execution_common.hpp (hunk 1 above)
Comment (chhwang) [reply src/include/execution_common.hpp:14]: Thank you, let's discuss this in the next version.

**Pull request merged**